### PR TITLE
Support for `HeteroData` in `Dataloader`

### PR DESCRIPTION
**Pull request by rusty1s**

### CODE REVIEW

**Comment by rusty1s** [comment]
test

**Comment by rusty1s** [comment]
This PR is now ready for review @RexYing. Let me know if there is anything unclear or what part might need more documentation. Overall, the following parts have changed:

* The `Batch` class can now dynamically inherit from `Data` or `HeteroData`, dependent on which class is found in the `data_list` to collate.
* I out-sourced the functionality of collating data lists and separating mini-batches to single data objects into their own respective classes. This helps us to re-use these functionalities for collating inside `InMemoryDataset` as well.
* Collating and separating is now performed recursively. This allows us to collate/separate dictionaries as well as lists, which was not possible before.

File: torch_geometric/data/collate.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    if isinstance(elem, Mapping):
+        # Recursively collate elements of dictionaries.
+        value_dict, slice_dict, inc_dict = {}, {}, {}
+        for key in elem.keys():
+            value_dict[key], slice_dict[key], inc_dict[key] = _collate(
+                key, [v[key] for v in values], data_list, stores, increment)
+        return value_dict, slice_dict, inc_dict
+
+    elif isinstance(elem, Sequence) and not isinstance(elem, str):
```
Comment (RexYing) [inline torch_geometric/data/collate.py:108]: Is sequence a necessary component? For hetero graph, is the attr usually dict (Mapping), or tensor? Is sequence here to accommodate previous implementation of collate?

File: torch_geometric/data/batch.py [hunk 2]
```diff
@@ -237,31 +115,35 @@ def index_select(self, idx: IndexType) -> List[Data]:
...
-            return self.get_example(idx)
+    def __getitem__(self, idx: Union[int, np.integer, str, IndexType]) -> Any:
+        if (isinstance(idx, (int, np.integer))
+                or (isinstance(idx, Tensor) and idx.dim() == 0)
+                or (isinstance(idx, np.ndarray) and np.isscalar(idx))):
+            return self.get(idx)
+        elif isinstance(idx, str):
+            return super().__getitem__(idx)
         else:
             return self.index_select(idx)
```
[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/data/batch.py:128]: I don't think that's possible as neither `self.get`, `super().__getitem__` or `self.index_select` utilize any recursion.

File: torch_geometric/data/collate.py (hunk 1 above)
Comment (rusty1s) [reply torch_geometric/data/collate.py:108]: `Sequence` could not be handled before, but I added it in as an enhancement independent of `HeteroData`. In particular, this is useful when you work on temporal graphs, where you might want to store features in lists, e.g.:
```python
data.edge_index = [edge_index_1, ..., edge_index_T]
```

**Review by RexYing — COMMENTED** [review]
i think the general code implementation makes sense. I wrote similar things for deepsnap before and it's indeed complex. had a couple questions. The test runs on my machine too. Will continue with more testing tmr.

File: torch_geometric/data/collate.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        return value_list, slice_list, inc_list
+
+    elif isinstance(elem, Tensor):
+        # Concatenate a list of `torch.Tensor` along the `cat_dim`.
+        # NOTE: We need to take care of incrementing elements appropriately.
+        cat_dim = data_list[0].__cat_dim__(key, elem, stores[0])
+        if cat_dim is None or elem.dim() == 0:
+            values = [value.unsqueeze(0) for value in values]
+        slices = cumsum([value.size(cat_dim or 0) for value in values])
+        if increment:
```
[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/data/collate.py:126]: `slices` is just to separate collated back tensors to original ones, e.g., it holds a compressed tensor holding the starting and ending points to "slice" the tensor to obtain original values. You are correct that both `inc` and `slice` use `cumsum` as they both hold the cumulated sum of previous values. I will try to make that more clear in a comment.

**Comment by rusty1s** [comment]
I added more tests for mini-batching, in particular, tests for verifying that both heterogeneous batching and recursive batching with lists and dictionaries work as expected.

File: torch_geometric/data/collate.py (hunk 1 above)
Comment (RexYing) [reply torch_geometric/data/collate.py:119]: i see. might be good to document this functionality too. Maybe at the function signature.

File: torch_geometric/data/separate.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+from torch import Tensor
+from torch_sparse import SparseTensor
+
+from torch_geometric.data.data import BaseData
+from torch_geometric.data.storage import BaseStorage
+
+
+def separate(cls, batch: BaseData, idx: int, slice_dict: Any,
+             inc_dict: Any = None, decrement: bool = True) -> BaseData:
```
Comment (RexYing) [inline torch_geometric/data/separate.py:13]: docstring here?

File: test/data/test_batch.py [hunk 5]
```diff
@@ -85,8 +112,110 @@ def __cat_dim__(self, key, item):
...
+    assert out1[e].edge_index.tolist() == data1[e].edge_index.tolist()
+    assert torch.allclose(out1[e].edge_attr, data1[e].edge_attr)
+
+    out2 = batch[1]
+    assert len(out2) == 3
+    assert out2.num_nodes == 150
+    assert torch.allclose(out2['p'].x, data2['p'].x)
+    assert torch.allclose(out2['a'].x, data2['a'].x)
+    assert out2[e].edge_index.tolist() == data2[e].edge_index.tolist()
+    assert torch.allclose(out2[e].edge_attr, data2[e].edge_attr)
```
Comment (RexYing) [inline test/data/test_batch.py:221]: just to make sure, if batch[1] edge index is correct, does it mean that inc will guarantee to work correctly?

File: test/data/test_batch.py [hunk 6]
```diff
@@ -85,8 +112,110 @@ def __cat_dim__(self, key, item):
...
+            '1': torch.randn(10, 32),
+            '2': torch.randn(20, 48)
+        }, edge_index=[get_edge_index(30, 30, 50),
+                       get_edge_index(30, 30, 70)], num_nodes=30)
+
+    data2 = Data(
+        x={
+            '1': torch.randn(20, 32),
+            '2': torch.randn(40, 48)
+        }, edge_index=[get_edge_index(60, 60, 80),
```
Comment (RexYing) [inline test/data/test_batch.py:135]: do we want to add a multiple edge type case as well?

**Review by RexYing — COMMENTED** [review]
tests look good to me. only minor questions

File: test/data/test_batch.py (hunk 6 above)
Comment (rusty1s) [reply test/data/test_batch.py:135]: Added :)

File: test/data/test_batch.py (hunk 5 above)
Comment (rusty1s) [reply test/data/test_batch.py:221]: You are right. We need at least three elements.

Comment (rusty1s) [reply test/data/test_batch.py:221]: Added this in :)

**Pull request merged**